### PR TITLE
Allow Dgraph to run badger in memory

### PR DESCRIPTION
**Pull request by harshil-goel**

### CODE REVIEW

File: dgraph/cmd/alpha/run.go [hunk 1]
```diff
@@ -683,7 +681,7 @@ func run() {
 
 	raft := z.NewSuperFlag(Alpha.Conf.GetString("raft")).MergeAndCheckDefault(worker.RaftDefaults)
 	x.WorkerConfig = x.WorkerOptions{
-		TmpDir:              Alpha.Conf.GetString("tmp"),
+		TmpDir:              "./t",
```
Comment (mangalaman93) [inline dgraph/cmd/alpha/run.go:684]: do we need this?

File: dgraph/cmd/alpha/run.go [hunk 2]
```diff
@@ -700,7 +698,9 @@ func run() {
 		Badger:              bopts,
 	}
 	x.WorkerConfig.Parse(Alpha.Conf)
+	fmt.Println("WORKER CONFIG ============", x.WorkerConfig)
```
Comment (mangalaman93) [inline dgraph/cmd/alpha/run.go:701]: will this print correctly? Could you copy an example here?

File: dgraph/cmd/alpha/run.go [hunk 3]
```diff
@@ -700,7 +698,9 @@ func run() {
...
 
+	worker.SetConfiguration(&opts)
```
Comment (mangalaman93) [inline dgraph/cmd/alpha/run.go:703]: can you add a comment to why this needs to be moved down here?

File: raftwal/log.go [hunk 4]
```diff
@@ -114,7 +114,7 @@ func openLogFile(dir string, fid int64) (*logFile, error) {
 			Dir:                           dir,
 			EncryptionKey:                 encryptionKey,
 			EncryptionKeyRotationDuration: 10 * 24 * time.Hour,
-			InMemory:                      false,
+			InMemory:                      true,
```
Comment (mangalaman93) [inline raftwal/log.go:117]: we are doing this permanent?

File: worker/config.go [hunk 5]
```diff
@@ -84,8 +84,6 @@ func (opt *Options) validate() {
 	x.Check(err)
 	td, err := filepath.Abs(x.WorkerConfig.TmpDir)
 	x.Check(err)
-	x.AssertTruef(pd != wd,
```
Comment (mangalaman93) [inline worker/config.go:87]: we should add a special condition for in memory.

File: worker/server_state.go [hunk 6]
```diff
@@ -118,19 +119,29 @@ func (s *ServerState) initStorage() {
...
+		}
 		s.WALstore, err = raftwal.InitEncrypted(Config.WALDir, x.WorkerConfig.EncryptionKey)
 		x.Check(err)
 	}
 	{
 		// Postings directory
 		// All the writes to posting store should be synchronous. We use batched writers
 		// for posting lists, so the cost of sync writes is amortized.
-		x.Check(os.MkdirAll(Config.PostingDir, 0700))
+		fmt.Println("Here", Config.PostingDir)
```
Comment (mangalaman93) [inline worker/server_state.go:132]: remove this

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
What are our plans for this PR? Do we want to add a separate option in Dgraph for in memory badger?

File: dgraph/cmd/alpha/run.go [hunk 7]
```diff
@@ -671,8 +671,6 @@ func run() {
 		os.Exit(1)
 	}
 
-	worker.SetConfiguration(&opts)
```
Comment (harshil-goel) [inline dgraph/cmd/alpha/run.go:674]: Why?

**Review by mangalaman93 — CHANGES_REQUESTED** [review]
We should add a test to ensure that we don't break it in future.

File: worker/server_state.go [hunk 8]
```diff
@@ -118,19 +118,25 @@ func (s *ServerState) initStorage() {
 
 	{
 		// Write Ahead Log directory
-		x.Checkf(os.MkdirAll(Config.WALDir, 0700), "Error while creating WAL dir.")
+		if Config.WALDir != "" {
```
Comment (sanjayk-github-dev) [inline worker/server_state.go:121]: @harshil-goel just trying to understand the changes. 

Is it the case that when badger is run in-memory, WALDir and PostingDir will be empty? And raftwal.InitEncrypted() handles this case properly?

**Pull request closed without merging**